lib/Parse/ParseDecl.cpp

```diff
@@ -3236,6 +3309,25 @@ bool Parser::parseDeclList(SourceLoc LBLoc, SourceLoc &RBLoc,
   return !RBLoc.isValid();
 }
 
+bool Parser::canDelayBodyParsing() {
+  if (SF.Kind == SourceFileKind::REPL)
+    return false;
+  if (SF.shouldCollectToken())
+    return false;
+  if (SF.shouldBuildSyntaxTree())
+    return false;
+  if (InPoundLineEnvironment)
```

Comment:
Huh. Is this because #sourceLocation is implemented via the virtual file mechanism, and we can't go back and parse those later?

Reply:
yeah, it's also because recovering parser status later for #sourceLocaion is not-trivial and may not worth it.